File: apps/worker/src/routes/customers.ts
import { Hono } from 'hono'
import type { Env } from '../index'
import { getUser } from '../middleware/auth'
import { getSupabase } from '../lib/supabase'

const app = new Hono<{ Bindings: Env }>()

async function tenantId(user: Record<string, unknown>, env: Env): Promise<string | null> {
  const sb = getSupabase(env)
  const { data } = await sb.from('nano_tenants').select('id').eq('owner_phone', user.phone_number).maybeSingle()
  return data?.id ?? null
}

app.get('/', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const limit = Math.min(parseInt(c.req.query('limit') ?? '20'), 100)
  const offset = parseInt(c.req.query('offset') ?? '0')
  const sb = getSupabase(c.env)
  const { data } = await sb.from('nano_customers').select('*').eq('tenant_id', tid).order('created_at', { ascending: false }).range(offset, offset + limit - 1)
  return c.json({ data: data ?? [], limit, offset })
})

app.post('/', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const body = await c.req.json<Record<string, unknown>>()
  const payload = Object.fromEntries(Object.entries(body).filter(([, v]) => v != null))
  payload.tenant_id = tid
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('nano_customers').insert(payload).select().single()
  if (error) return c.json({ error: error.message }, 503)
  return c.json(data)
})

app.get('/:id', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const sb = getSupabase(c.env)
  const { data } = await sb.from('nano_customers').select('*').eq('id', c.req.param('id')).eq('tenant_id', tid).maybeSingle()
  if (!data) return c.json({ error: 'Customer not found' }, 404)
  return c.json(data)
})

app.patch('/:id', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const body = await c.req.json<Record<string, unknown>>()
  const updates = Object.fromEntries(Object.entries(body).filter(([, v]) => v != null))
  if (!Object.keys(updates).length) return c.json({ error: 'No fields to update' }, 400)
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('nano_customers').update(updates).eq('id', c.req.param('id')).eq('tenant_id', tid).select().single()
  if (error) return c.json({ error: error.message }, 503)
  if (!data) return c.json({ error: 'Customer not found' }, 404)
  return c.json(data)
})

app.delete('/:id', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('nano_customers').delete().eq('id', c.req.param('id')).eq('tenant_id', tid).select()
  if (error) return c.json({ error: error.message }, 503)
  if (!data?.length) return c.json({ error: 'Customer not found' }, 404)
  return c.json({ deleted: true })
})

export default app

File: apps/api/app/routers/customers.py
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from app.auth import get_current_user
from app.deps import optional_supabase

router = APIRouter(prefix="/customers", tags=["customers"])


class CustomerBody(BaseModel):
    name: str | None = None
    email: str | None = None
    phone: str | None = None
    metadata: dict | None = None


async def _tenant_id(user: dict, db: Any) -> str:
    row = (
        db.table("nano_tenants")
        .select("id")
        .eq("owner_phone", user["phone_number"])
        .maybe_single()
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Tenant not found")
    return row.data["id"]


@router.get("")
async def list_customers(
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    row = (
        db.table("nano_customers")
        .select("*")
        .eq("tenant_id", tid)
        .order("created_at", desc=True)
        .range(offset, offset + limit - 1)
        .execute()
    )
    return {"data": row.data, "limit": limit, "offset": offset}


@router.post("")
async def create_customer(
    body: CustomerBody,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    payload = {k: v for k, v in body.model_dump().items() if v is not None}
    payload["tenant_id"] = tid
    row = db.table("nano_customers").insert(payload).execute()
    return row.data[0]


@router.get("/{customer_id}")
async def get_customer(
    customer_id: str,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    row = (
        db.table("nano_customers")
        .select("*")
        .eq("id", customer_id)
        .eq("tenant_id", tid)
        .maybe_single()
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Customer not found")
    return row.data


@router.patch("/{customer_id}")
async def update_customer(
    customer_id: str,
    body: CustomerBody,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    updates = {k: v for k, v in body.model_dump().items() if v is not None}
    if not updates:
        raise HTTPException(400, "No fields to update")
    row = (
        db.table("nano_customers")
        .update(updates)
        .eq("id", customer_id)
        .eq("tenant_id", tid)
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Customer not found")
    return row.data[0]


@router.delete("/{customer_id}")
async def delete_customer(
    customer_id: str,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    row = (
        db.table("nano_customers")
        .delete()
        .eq("id", customer_id)
        .eq("tenant_id", tid)
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Customer not found")
    return {"deleted": True}
